Interactivity API: Improvements to the experimental full-page navigation

michalczaplinski opened the pull request:
This PR brings several improvements to experimental full-page navigation. 

Fixes #63880

-----

### 1. Remove `src` attributes from prefetched scripts

We have to remove the `src` attributes of the prefetched scripts because otherwise the contents are ignored. Reported by @luisherranz in https://github.com/WordPress/gutenberg/issues/63880#issuecomment-2262808259.

### 2. Use `.textContent` instead of `.innerText` to set `<script>` contents

Prefetched scripts that are inserted in the the `<head>` have `<br>` in the DOM:

<img width="500" alt="Screenshot 2024-07-29 at 16 49 54" src="https://github.com/user-attachments/assets/5d1b33e3-b6e0-4575-9372-fd52f4aa5b3f">

The article from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText) about `innerText` explains why we see this problem:

> As a setter this will replace the element's children with the given value, converting any line breaks into [<br>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/br) elements.

### 3. `populateInitialData()` with state from the server

We have to populate the client-side state with the state coming from the server (as mentioned in [this comment](https://github.com/WordPress/gutenberg/issues/63880#issuecomment-2256300053):

> We populate the client-side state with the state from the server on the initial load.
> We also re-populate the state when doing region-based navigation. But we don't do the equivalent when doing full page navigations.

### 4. Wait for `load` event of the `script` element

Finally, there is another issue:

Scripts that are added to the DOM with `type="module"` are deferred by default! This means that if we add a script to the `<head>` the way we currently do it is not guaranteed to have executed before the HTML of the page is updated. 

Notice how the DOM had been updated before the `onload` handler fires. The error at the end that is caused by it because the JS store that contains the `callbacks.initTriggerButton` has not yet loaded on the page.

https://github.com/user-attachments/assets/ef25c22e-8877-4d1e-97c7-5ac8b05c1eaa

The solution proposed in this PR follows [this comment](https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2271003468):
- For prefetching create `<link>` elements with `modulepreload` instead of fetching the modules ourselves.
- Upon navigation, use `import()` to evaluate the scripts. 


## Testing instructions

You can test this PR manually by using blocks that use the Interactivity API and enabling the client-side navigation in GB settings.

Some things to watch out for:
- Watch out for any client-side console errors.
- Ensure the JS assets are preloaded correctly (watch the Network tab of your browser's devtools). They should be preloaded on hovering over a link.
- Ensure that the cache-busting works correctly (as referred to in [this comment](https://github.com/WordPress/gutenberg/pull/64067#issuecomment-2320781293))
- Ensure that upon hovering over a link, we only pre-load modules, NOT all scripts.

### Example test 

1. Make sure to enable the "full page client-side navigation" experiment in GB settings
2. Create a post/page with an Image block using the "Lightbox" feature. Let's call it an `Image test` post.
3. Navigate to your **Blog Archive** page.
4. Hover over the link to the `Image test` post.
5. **TEST:** Ensure that the assets are preloaded correctly.
6. Click on the link to the `Image test` post.
7. **TEST:** Ensure that the post is navigated to using full-page client-side navigation.
8. **TEST:** Click on the image to open the Lightbox. **Ensure that it can be opened and closed without issue**.
9.  Navigate back to the Blog Archive page. 
10. **TEST:** Ensure that the navigation was done using full-page client-side navigation.

----

Several follow-ups should be done:

1. Figure out the way to only load the scripts for the new & compatible interactive blocks on the page. Previously mentioned in https://github.com/WordPress/gutenberg/issues/63880#issuecomment-2263605123 & https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2106176106

2. Refactor the `fetchHeadAssets()` function so that it's not called inside of `regionsToVdom`.

3. Handle CSS asset loading (not done in this PR): 

   1. Handle properly the cache busting hash in the URLs of the block styles. [See this comment](https://github.com/WordPress/gutenberg/pull/64067#issuecomment-2317367518)
   2. Moving the styles from a file to `<style>` tags messes up with how the URLs are interpreted in CSS's  `url()` function: https://github.com/WordPress/gutenberg/pull/64067#discussion_r1738591715

4. Occasionally the CSS for the page after a navigation is messed up (see video). I'm not sure if this is due to missing CSS or or file order or something else.

   https://github.com/user-attachments/assets/586406c7-5634-4da5-9508-786b651870fa

michalczaplinski committed:
fix: Add initial data population in interactivity router

michalczaplinski commented:
I believe that this fix should work after https://github.com/WordPress/gutenberg/pull/62734 is merged.

michalczaplinski committed:
chore: Update element.innerText to element.textContent in head.ts file

michalczaplinski committed:
feat: Exclude src and href attributes when copying head element attributes

michalczaplinski committed:
chore: Populate initial data in interactivity router

michalczaplinski committed:
chore: Move Populate initial data in interactivity router

michalczaplinski committed:
Wait for `load` event of script element before returning from `fetchHeadAssets()` function

luisherranz commented:
> I believe that this fix should work after https://github.com/WordPress/gutenberg/pull/62734 is merged.

What functionality or change from the other pull request does this one need to work?

michalczaplinski commented:
> What functionality or change from the other pull request does this one need to work?

Sorry, this comment is outdated and I've removed it now. I made it before updating this PR . We don't need any specific functionality from that PR.

luisherranz commented:
I added a comment [here](https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2271003468) about the general strategy of preloading and waiting for resources to load.

michalczaplinski commented:
I've just noticed that this:

```js
// wait for the `load` event to fire before appending the element
return new Promise( ( resolve, reject ) => {
  element.onload = resolve;
  element.onerror = reject;
} );
```

actually breaks the full-page navigation. We can't rely on the `onload` event I guess. I'll take a look shortly.

michalczaplinski commented:
> I've just noticed that this:
> 
> ```js
> // wait for the `load` event to fire before appending the element
> return new Promise( ( resolve, reject ) => {
>   element.onload = resolve;
>   element.onerror = reject;
> } );
> ```
> 
> actually breaks the full-page navigation. We can't rely on the `onload` event I guess. I'll take a look shortly.

Just coming back to this. The reason I was seeing breakage is because the `load` event never fires on inline `<script>` elements. So something like this does not work:

```js
     const script = document.createElement('script');
     script.textContent = 'console.log("hi");'    
     script.onload = () => { console.log("loaded")}; // This will log
     document.body.append(script);
```

I'm trying a different approach now: 

1. Instead of loading the content with `fetch()` and inlining it into `<script>` and `<style>` elements, fetch it with `<link rel="modulepreload">`. For styles, we can use `rel="preload" as="style"`.
2. In the `updateHead()` return a Promise waiting for the `load` of the script.

luisherranz commented:
> - In the `updateHead()` return a Promise waiting for the `load` of the script.

As I mentioned [here](https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2271003468), you don't need to add an actual script and use the `load` event for the scripts; you can use dynamic imports since all the scripts should be modules.

michalczaplinski committed:
feat: Update head tags to improve prefetching of scripts and stylesheets

This commit modifies the `updateHead` function in `head.ts` to improve support for lazy loading of scripts and stylesheets. It preloades the script modules using `modulepreload`, imports the necessary scripts using dynamic imports and adds the `preload` link elements for stylesheets.

michalczaplinski commented on packages/interactivity-router/src/index.ts:
```diff
@@ -54,7 +54,10 @@ const navigationMode: 'regionBased' | 'fullPage' =
 // The cache of visited and prefetched pages, stylesheets and scripts.
 const pages = new Map< string, Promise< Page | false > >();
-const headElements = new Map< string, { tag: HTMLElement; text: string } >();
+export const headElements = new Map<
+	string,
+	{ tag: HTMLElement; text?: string }
+>();
```
I've exported the `headElements` so that we don't have to pass it as an argument to `fetchHeadAssets()`.

michalczaplinski commented on packages/interactivity-router/src/head.ts:
```diff
@@ -29,6 +34,14 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+		[ ...headElements.entries() ]
+			.filter( ( [ , { tag } ] ) => tag.nodeName === 'SCRIPT' )
+			.map( async ( [ url ] ) => {
+				await import( /* webpackIgnore: true */ url );
```
We need the webpack comment here. Otherwise, webpack will try to code-split here into a new chunk and it [does not permit](https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import) using fully dynamic paths.

michalczaplinski commented:
> As I mentioned https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2271003468, you don't need to add an actual script and use the load event for the scripts; you can use dynamic imports since all the scripts should be modules.

Yup, this is what I ended up doing. I think there's nothing wrong with adding the scripts and waiting for the `load` approach, though. 

---

I think this is ready for review now. 

There are 2 issues that I've noticed that should be fixed in follow-ups:

1. Do not load the scripts that have already been preloaded if they only differ by `?ver=<some-number>` query param. I initially thought that this should be fixed when using the production build (`npm run build`) but it's not the case.


   https://github.com/user-attachments/assets/66dbd9ea-d36d-4665-a88b-688ca6e03493



2. Occasionally the CSS for the page after a navigation is messed up (see video). I'm not sure if this is due to missing CSS or or file order or something else.

   https://github.com/user-attachments/assets/586406c7-5634-4da5-9508-786b651870fa

gziolo commented:
> I initially thought that this should be fixed when using the production build (npm run build) but it's not the case.

You should test it with `SCRIPT_DEBUG` set to `false` in your `wp-config.php` file.

michalczaplinski commented:
> > I initially thought that this should be fixed when using the production build (npm run build) but it's not the case.
> 
> You should test it with `SCRIPT_DEBUG` set to `false` in your `wp-config.php` file.

Nice, I didn't think about it, thanks! 🙂

michalczaplinski commented:
ok, I've just tried it and it looks like setting `SCRIPT_DEBUG` to `false` does not make any difference, unfortunately.

gziolo commented:
> ok, I've just tried it and it looks like setting `SCRIPT_DEBUG` to `false` does not make any difference, unfortunately.

This is the reason it uses `time()` for interactivity router script:

https://github.com/WordPress/gutenberg/blob/0e88cdae4586a1b071bf135b08bb33c2c0c3880d/lib/interactivity-api.php#L13

I don't know where the `style.css` comes from, but there must be something similar present in the code.

michalczaplinski commented:
Thanks @gziolo 

I figured out that `GUTENBERG_VERSION` is only available in production. In dev, it's unset.

https://github.com/WordPress/gutenberg/blob/fb5fa24ae1ef25f2f9dfb83a87d7a1b369541912/bin/generate-gutenberg-php.php#L22-L23

For **production**, there is no problem in that case. `ver === GUTENBERG_VERSION` 

In **development**, we should check if the experiment is enabled. And use the `time()` only if the experiment is disabled.

michalczaplinski committed:
Do not load interactivity script modules in development mode when full page navigation is enabled

michalczaplinski committed:
Format interactivity-api.php

michalczaplinski commented:
> In development, we should check if the full-page navigation experiment is enabled. And use the `time()` only if the full-page navigation experiment is disabled.

I've now added this. 

We should look into doing the same for indivudual block's stylesheets:

<img width="532" alt="Screenshot 2024-08-29 at 12 19 32" src="https://github.com/user-attachments/assets/7526995a-38b2-4f10-b89f-c141823e2479">

But this can be done in a follow-up

luisherranz commented:
> In **development**, we should check if the full-page navigation experiment is enabled. And use the `time()` only if the experiment is disabled.

Won't that cause problems for people developing in Gutenberg with that option enabled?

michalczaplinski commented:
> Won't that cause problems for people developing in Gutenberg with that option enabled?

I think that if you are developing with the full-page navigation enabled, then you DO want to disable the cache busting. Otherwise, full-page navigation does not work properly. This is how I did it:

https://github.com/WordPress/gutenberg/blob/bcb3ff8ea4aaba550529e55e50c0291735133e7e/lib/interactivity-api.php#L17-L25

Am I missing something?

luisherranz commented:
> I think that if you are developing with the full-page navigation enabled, then you DO want to disable the cache busting. Otherwise, full-page navigation does not work properly.

Won't disabling cache busting cause problems for people who are developing? Is it possible that some people developing have activated the full-page navigation experiment and, without knowing why, suddenly their development environment is broken?

DAreRodz commented on packages/interactivity-router/src/index.ts:
```diff
@@ -109,17 +112,17 @@ const regionsToVdom: RegionsToVdom = async ( dom, { vdom } = {} ) => {
 // Render all interactive regions contained in the given page.
 const renderRegions = ( page: Page ) => {
-	batch( () => {
+	batch( async () => {
```
Have you tested `batch()` works with asynchronous functions? 🤔 

I think there was a reason for calling `render()` and `populateServerData()` together inside the `batch()` callback, but I don't remember (maybe avoid triggering any updates before render?).

michalczaplinski replied:
I have not tested it explicitly! Thanks for pointing it out 🙂 Let me test and get back to you here.

michalczaplinski replied:
It should be completely safe to call `batch()` with an async function: 

https://github.com/preactjs/signals/blob/e04671469e9272de356109170b2e429db49db2f0/packages/core/src/index.ts#L98-L102

It's just a wrapper that internally calls the passed function without any modification. It just does some bookkeeping to flush the signals once the function returns.

DAreRodz commented on packages/interactivity-router/src/index.ts:
```diff
@@ -109,17 +112,17 @@ const regionsToVdom: RegionsToVdom = async ( dom, { vdom } = {} ) => {
 			if ( navigationMode === 'fullPage' ) {
 				// Once this code is tested and more mature, the head should be updated for region based navigation as well.
-				updateHead( page.head );
+				await updateHead( page.head );
```
Do you think it's possible to move this `await` outside of the `renderRegions()` logic?

I feel `renderRegions()` should be synchronous. 😅 Maybe we can move the `updateHead()` call right before running `renderRegions()`, [here](https://github.com/WordPress/gutenberg/blob/bcb3ff8ea4aaba550529e55e50c0291735133e7e/packages/interactivity-router/src/index.ts#L162) and [here](https://github.com/WordPress/gutenberg/blob/bcb3ff8ea4aaba550529e55e50c0291735133e7e/packages/interactivity-router/src/index.ts#L307).

This way, we wouldn't need an async function inside `renderRegions()`.

michalczaplinski replied:
Good thinking! I will try the other solution - we can preload the scripts in a
separate function like you suggested in the other comment.

michalczaplinski replied:
OK, that does not quite make sense. I've just tried making `renderRegions()` sync but it's not really possible as we need to wait for the assets to be preloaded before updating the HTML.

michalczaplinski replied:
I gave it some thought and tried to make the `renderRegions()` sync but I think
that it's not the right pattern nor is it necessary. 

We always **have to** wait for the new JS to load before rendering the new
HTML. By "new HTML", I mean here HTML of a new page using client-side
full-page navigation or a new region using render-based navigation.

If we don't await the JS that might have been loaded on the new page, we run
into the very same issue that this PR is trying to fix:
https://github.com/WordPress/gutenberg/issues/63880 😃

DAreRodz commented on packages/interactivity-router/src/head.ts:
```diff
@@ -29,6 +34,14 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
 		}
 	}
 
+	await Promise.all(
```
Another option we could try is to move this code to its own function. This way, `updateHead()` would become synchronous again, and we wouldn't have to use `await` inside `renderRegions()`.

michalczaplinski replied:
Yeah, that might work! Let's try 🙂

DAreRodz commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +54,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
-	}
+			}
+
+			const headElement = headElements.get( href );
+			const styleElement = doc.createElement( 'style' );
+			styleElement.textContent = headElement.text;
```
We could try something similar to what you added for script modules, i.e., adding `<link>` tags with the `preload` attribute.

There is an issue with moving the stylesheet content to `<style>` tags: the base URL for relative URLs would change. They're supposed to use [the stylesheet URL as the base URL](https://www.w3.org/TR/CSS1/#:~:text=Partial%20URLs%20are%20interpreted%20relative%20to%20the%20source%20of%20the%20style%20sheet%2C%20not%20relative%20to%20the%20document), but that would change to the document URL if we do this. ☝️

michalczaplinski replied:
Good catch there 🙂 .

I didn't handle the styles in this PR because it already has quite a few changes that are only loosely related. I'd prefer to tackle styles in a separate follow-up PR. I'll note it down in PR description.

DAreRodz reviewed (commented):
I've left a couple of comments and suggestions. 🙂

michalczaplinski commented:
@gziolo I see that the `*asset.php` file is [optional](https://core.trac.wordpress.org/ticket/60460). Thus, we could check if the file is present: 
- If present, then we use the hash from that asset file as the version.
- If the `*.asset.php` file is missing we don't use the `version`

The above logic would only apply **in development** and **only if the full-page navigation experiment is enabled**.

gziolo commented:
@michalczaplinski, sure that might work. The asset file is always created by the build process in the Gutenberg plugin. You can add a fallback as it's probably handled that way in other places, too.

michalczaplinski committed:
Update interactivity script module registration to use version from asset files

- Added logic to retrieve version information from `index.min.asset.php` and `router.min.asset.php` files.
- Updated `wp_register_script_module` calls to use the retrieved version instead of the default version when full-page navigation is not enabled.

michalczaplinski committed:
empty commit

michalczaplinski committed:
Rename populateInitialData to populateServerData

michalczaplinski committed:
remove populateServerData

michalczaplinski committed:
try: remove the webpack comment

michalczaplinski committed:
try: remove the await import()

michalczaplinski committed:
bring back the async import

michalczaplinski committed:
Move headElements to head.ts

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -29,6 +37,14 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+		[ ...headElements.entries() ]
+			.filter( ( [ , { tag } ] ) => tag.nodeName === 'SCRIPT' )
+			.map( async ( [ url ] ) => {
+				await import( url );
```
This line was causing problems with DEWP, is that correct? That suggests that webpack is trying to handle this import which is effectively a wildcard. That "URL" could be anything.

Webpack may be trying to bundle up whatever it can find.

This should probably have a `/* webpackIgnore: true */` [magic comment](https://webpack.js.org/api/module-methods/#magic-comments). Webpack should not be handling this import in any way.

michalczaplinski replied:
Nope, this was not the issue. The issue was that there was a cyclical dependency between `index.js` and `head.ts ` in the router package. That was fixed (worked around, really) in https://github.com/WordPress/gutenberg/pull/64067/commits/c7e70ddcf62371a5309626487b672246d42165b7 by moving the `headElements` variable to the `head.ts` file.

But you're right that we need the webpack comment here! I've already had it before but temporarily removed it in https://github.com/WordPress/gutenberg/pull/64067/commits/62e527e1769162894881eba76fc33bd96e3fa7eb when trying to debug the problem with cyclical dependencies.

michalczaplinski committed:
Revert "try: remove the webpack comment"

This reverts commit 62e527e1769162894881eba76fc33bd96e3fa7eb.

michalczaplinski committed:
default_version => router_version

michalczaplinski commented:
I believe this should be good to be re-reviewed again @gziolo @DAreRodz @sirreal 🙇‍♂️

gziolo commented:
What is the best way to test these changes? How much confidence can we put in the existing e2e tests coverage?

michalczaplinski commented:
>  What is the best way to test these changes? How much confidence can we put in the existing e2e tests coverage?

Apologies, I've now added better testing instructions in the PR description.

The existing e2e tests do not cover the scenarios in this PR. I have started working on additional e2e tests in https://github.com/WordPress/gutenberg/pull/63268 

I'm happy to continue working on them as an immediate follow-up. We'll need them before full-page client-side navigation is moved out of GB experiment. For now, as full-page client-side navigation is still experimental, I'm not worried about the impact of this PR. The current implementation has been somewhat broken for months (which this PR attempts to fix 🙂).

sirreal commented on lib/interactivity-api.php:
```diff
@@ -14,13 +14,24 @@ function gutenberg_reregister_interactivity_script_modules() {
 	wp_deregister_script_module( '@wordpress/interactivity' );
 	wp_deregister_script_module( '@wordpress/interactivity-router' );
 
+	$experiments                  = get_option( 'gutenberg-experiments' );
+	$full_page_navigation_enabled = isset( $experiments['gutenberg-full-page-client-side-navigation'] );
+
+	$index_path    = gutenberg_dir_path() . 'build/interactivity/index.min.asset.php';
+	$index_asset   = file_exists( $index_path ) ? require $index_path : null;
+	$index_version = isset( $index_asset['version'] ) ? $index_asset['version'] : $default_version;
+
```
This needs conflict fixes from #65460. What are the changes to this file for? You may be able to remove all the changes here 🤞 

It looks like this is intended to remove a version from the `@wordpress/interactivity-router` package if full page navigation is enabled. I suspect that is because a time-based default was used before which would make the URL for the module not match across page loads. With #65460, a stable version hash should be used from the generated asset file. If the module changes, the hash will change, but that should only happen during active development and would require a reload of the module anyways.

michalczaplinski replied:
Thank you Jon! 

> It looks like this is intended to remove a version from the @wordpress/interactivity-router package if full page navigation is enabled. I suspect that is because a time-based default was used before which would make the URL for the module not match across page loads.

That's exactly right. I'll rebase and check if it works as expected now.

michalczaplinski replied:
I just tested it and indeed it works fine after #65460 was merged 👍

michalczaplinski committed:
Remove the changes to interactivity-api.php

michalczaplinski committed:
Merge branch 'trunk' into fix/full-page-navigation-server-state

gziolo commented on packages/interactivity-router/src/index.ts:
```diff
@@ -170,13 +170,15 @@ window.addEventListener( 'popstate', async () => {
-		} );
-		await fetchHeadAssets( document, headElements );
+		[].map.call(
+			document.querySelectorAll( 'script[type="module"][src]' ),
```
That's intriguing, as it skips all scripts that aren't ES Modules moving forward. It never covered inline scripts, to there is no change regarding that.

What's the story for the plugins or themes that add custom JavaScript to the site? It will load on the full-page reload but not when doing the full-page client-side navigation. It would be good to start collecting the requirements for the feature to work correctly.

michalczaplinski replied:
Yes, that's intentional. This was outlined by @luisherranz in https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2106176106: 

> @michalczaplinski We shouldn't really load unknown scripts. We just did it to get it working with the WP movies demo.
>
> We should only load the module scripts that belong to the new blocks on the page or other module scripts that have declared their compatibility with full-page client-side navigation.

One issue with scripts was that they could depend on the [loading order](https://github.com/WordPress/gutenberg/pull/59707#pullrequestreview-1975274682). That's because scripts (without `async` or `deferred` attributes) are blocking. Modules, on the other hand, are deferred by default and users should be able to load them in any order.

gziolo replied:
Let's make sure it's documented in the code so it's clear that it's intentional to support only ES modules served from files.

michalczaplinski replied:
Gotcha. Done in 375d4eea85cbdc382206c6b12c401ed22f6811a7

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -29,6 +37,14 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
 	}
 
+	await Promise.all(
+		[ ...headElements.entries() ]
```
**This looks good**, but I always worry about these data structures because I think of Map as an unordered collection. I was worried about dependencies and ordering not being respected:

> The `entries()` method of Map instances returns a new map iterator object that contains the [key, value] pairs for each element in this map **in insertion order**.

[From MDN (bold mine)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries).

It seems like we're OK since entries returns elements in insertion order 👍

gziolo commented:
We need to validate whether this PR fixes https://github.com/WordPress/gutenberg/issues/63880. If not, then we should not need to backport this PR to WP 6.7, so the corresponding label will have to be removed.

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+
+			const headElement = headElements.get( href );
+			const styleElement = doc.createElement( 'style' );
+			styleElement.textContent = headElement.text;
```
Is it also necessary to store the `text`? Couldn't `headElements` be `Map< string, HTMLElement >`, then this could become `styleElement.textContent = headElement.textContent`?

michalczaplinski replied:
Possibly yes...? :) 

I'd prefer to leave the improvements to style loading for a future PR because this one has already ballooned beyond its original scope. More than happy to add it in a follow-up.

sirreal replied:
It cannot. The style `Link` elements have `text` from the CSS stylesheet they download.

michalczaplinski replied:
You're right 🙂

DAreRodz committed:
Make `renderRegions` async

DAreRodz commented:
I have found an issue (maybe introduced after merging `trunk`?).

Following the steps listed in the description, I see this error when I navigate to the "Image test" page:

```
context.ts:62 Uncaught (in promise) TypeError: Invalid value used as weak map key
    at WeakMap.set (<anonymous>)
    at proxifyContext (context.ts:62:26)
    at directives.tsx:179:49
    at F (index.js:322:25)
    [...]
```

After a bit of debugging, the cause of the problem seems to be inside the `wp-context` directive. These refs lose their initial value somehow:

https://github.com/WordPress/gutenberg/blob/001241bbb3ff1b0c24c025815d7897ef353d2d74/packages/interactivity/src/directives.tsx#L149-L150

This is a quick fix we could use right after initialize the refs:

```js
if ( ! client.current ) {
    client.current = proxifyState( ns, {} );
}
if ( ! server.current ) {
    server.current = proxifyState( ns, {}, { readOnly: true } );
}
```

However, this is weird, and would require further investigation. It could be an issue in Preact, or the way it reconciles elements during rendering. I have no idea at this moment. 😅

DAreRodz commented:
@michalczaplinski, I've added minor changes to the `renderRegions()` function in https://github.com/WordPress/gutenberg/pull/64067/commits/a9469d0e19db5a639c6c0ab5580343282584983e. Just made the `renderRegions()` function async and wrapped only the required parts with `batch()`. 🙂

michalczaplinski commented:
@DAreRodz 

> I have found an issue (maybe introduced after merging trunk?).

I cannot reproduce this problem on my end. Are you sure this issue has been introduced in this PR?

michalczaplinski commented:
> We need to validate whether this PR fixes https://github.com/WordPress/gutenberg/issues/63880. If not, then we should not need to backport this PR to WP 6.7, so the corresponding label will have to be removed.

This PR does indeed fix it! It's mentioned in the description so the issue should auto-close.

michalczaplinski commented:
@gziolo @DAreRodz @sirreal 

I believe all the feedback has been addressed, and if there’s no major blocker, it would be great to give this PR a final review. 

In follow-up PRs, the following issues should be considered (with emphasis on 3. - CSS asset loading):

1. Figure out how to only load the scripts for the new & compatible interactive blocks on the page. Previously mentioned in https://github.com/WordPress/gutenberg/issues/63880#issuecomment-2263605123 & https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2106176106

2. Refactor the `fetchHeadAssets()` function so that it's not called inside of `regionsToVdom`.

3. Handle CSS asset loading: 

   1. Handle properly the cache busting hash in the URLs of the block styles. [See this comment](https://github.com/WordPress/gutenberg/pull/64067#issuecomment-2317367518)
   2. Moving the styles from a file to `<style>` tags messes up with how the URLs are interpreted in CSS's  `url()` function: https://github.com/WordPress/gutenberg/pull/64067#discussion_r1738591715
   3. Occasionally the CSS for the page after a navigation is messed up (see video). I'm not sure if this is due to missing CSS or or file order or something else.

   https://github.com/user-attachments/assets/586406c7-5634-4da5-9508-786b651870fa

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+		}
+	} );
+
+	const stylesheets = doc.querySelectorAll< HTMLScriptElement >(
```
Wrong element here:
```suggestion
	const stylesheets = doc.querySelectorAll< HTMLLinkElement >(
```

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
-				element.innerText = headElement.text;
-				for ( const attr of headElement.tag.attributes ) {
-					element.setAttribute( attr.name, attr.value );
+	Array.from( scripts ).forEach( ( script ) => {
```
Isn't that redundant?

```suggestion
	scripts.forEach( ( script ) => {
```

Stylistically (up to you), I've come to prefer `for...of` loops for this type of thing:

```js
for ( const script of doc.querySelectorAll< HTMLScriptElement >( 'script[type="module"][src]' ) ) {
  //…
}
```

michalczaplinski replied:
You're right! I didn't realize that NodeList had the `forEach()` method too!

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+
+	await Promise.all(
+		Array.from( stylesheets ).map( async ( tag ) => {
+			const href = tag.getAttribute( 'href' );
```
This is nullable… there should probable be an early return if there's no `href`:

```suggestion
			const href = tag.getAttribute( 'href' );
			if ( ! href ) {
				return;
			}
```

We need to remove `strict: false` from this project's tsconfig.

michalczaplinski replied:
Done in cf73a668fb5cae25511b51eaf45638164cb4f08e

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+			const href = tag.getAttribute( 'href' );
+			if ( ! headElements.has( href ) ) {
+				try {
+					const response = await fetch( href );
+					const text = await response.text();
```
This should check for `response.ok` and it would probably be good to check the content type response as well.

```js
if ( ! response.ok ) { return; }
if ( 'text/css' !== response.headers.get('Content-Type') ) { return; }
```

These _could_ throw an error to enter the catch and be logged to the console.

sirreal replied:
I'm rather new to this and lack context. Is there a reason to inject link tags instead of loading the style elements inline?

One reason I'm curious about this is that the browser will decide how many resources on the page to fetch in parallel, but I believe these fetch calls fetch all the resources in parallel without any limits. Ideally it seems like `link` tags would be used but I assume that was explored and discarded as an option.

michalczaplinski replied:
> This should check for `response.ok` and it would probably be good to check the content type response as well.

That totally makes sense, but could we also leave this for a follow-up PR? I didn't intend to change this behavior in this PR.

michalczaplinski replied:
> I'm rather new to this and lack context. Is there a reason to inject link tags instead of loading the style elements inline?

I'm not sure I follow here, sorry. We are injecting `<link>` tags to [prefetch modules](https://github.com/WordPress/gutenberg/blob/cf73a668fb5cae25511b51eaf45638164cb4f08e/packages/interactivity-router/src/head.ts#L77-L79). 

Or did you mean if there's a reason NOT to inject `link` tags for CSS assets?

sirreal replied:
This: why not use the same link tags for CSS assets?

michalczaplinski replied:
Yeah, we can do that! (https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2271003468)

I just wasn't planning on doing that in this iteration 🙂

sirreal reviewed (commented):
I've done some testing and am leaving initial feedback. I'll follow the complete testing instruction now.

Please add an entry to the package changelog.

---

One thing I know is missing and I don't know how to solve is related to the importmap.

- Only 1 importmap is allowed per page
- Once an importmap has been processed, others are ignored
- [Modifying the importmap does not work](https://github.com/WICG/import-maps)

  > Note that (like other <script> elements) modifying the contents of a <script 
    type="importmap"> after it's already inserted in the document will not work.

I'm not sure whether there's a solution at the technical level for this right now 😞 

This means that in a case where a full page navigation occurs and the importmap has additional assets, navigation may need to bail and let regular navigation take over.

sirreal commented on packages/interactivity-router/src/head.ts:
```diff
@@ -41,60 +57,59 @@ export const updateHead = async ( newHead: HTMLHeadElement[] ) => {
+		'link[rel=stylesheet]'
+	);
+
+	await Promise.all(
+		Array.from( stylesheets ).map( async ( tag ) => {
```
The possible fetch failure is one of the more fragile places here and it's handled. 

However, anything else that fails in these promises will cause the entire function to reject. I don't think that's desired (based on the `try/catch` on `fetch`). [`allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) may be better to use here, since we're not interested in the result and want to make a best effort here, then the function can continue whether or not _all_ the promises resolve.

michalczaplinski replied:
That's a good idea indeed, but could we do it in a follow-up PR? 

I didn't intend to change this behavior in this PR.

michalczaplinski committed:
Update TS type of the stylesheets variable

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

michalczaplinski committed:
Replace Array.from() with direct forEach() on NodeList in head.ts:

michalczaplinski committed:
Check if href is null

michalczaplinski commented:
The changes don't need to be backported to 6.7. They only affect the full-page navigation, which is only present as an experiment in GB.

The only part that affects the region-based navigation was the addition of batching using `batch()` by @DAreRodz inside of `renderRegions()` in https://github.com/WordPress/gutenberg/commit/a9469d0e19db5a639c6c0ab5580343282584983e as justfied by https://github.com/WordPress/gutenberg/pull/64067#issuecomment-2371529191. But that was more done out of precaution. It was not reported as a bug before.

michalczaplinski committed:
Clarify why we only prefetch script modules

sirreal reviewed (approved):
I'm comfortable landing this and handling other changes in follow-up work now that the backport label has been removed 👍

michalczaplinski committed:
Add changelog

michalczaplinski committed:
Merge branch 'trunk' into fix/full-page-navigation-server-state

michalczaplinski commented:
Thank you @sirreal ! 🙏

sirreal commented:
There are several follow-up items from this PR that I don't want to slip through the cracks. Are they tracked anywhere?

michalczaplinski commented:
Thanks for the vigilance @sirreal :) 

Yes, they're tracked in https://github.com/WordPress/gutenberg/issues/60951. In particular, see [this](https://github.com/WordPress/gutenberg/issues/60951#issuecomment-2413785152) latest comment.

mevanloon commented:
Regarding the CSS jumbling up, I have a strong suspicion this has to do with the way global styles are generated per block. Those get a suffix generated by `wp_unique_prefixed_id()`, and you'd get something like `wp-container-core-group-is-layout-1`. Though, this is not unique to the global site, just the current page. Those suffixes might collide and invalidate the CSS of an earlier page.

I encountered this issue with Swup (a JS framework that is remarkably similar to what the Interactivity API is doing), and 'fixed' it with this incredibly hacky hack:
```php
if(is_wp_version_compatible("6.5")) {
  add_filter('sanitize_title', function($title) {
    if(preg_match("/core-/", $title)) {
      $rand = rand(0, 1000);
      
      for($i=0;$i<$rand;$i++) {
        wp_unique_prefixed_id("wp-container-$title-is-layout-");
      }
    }
    return $title;
  });
}
```
